### PR TITLE
Add DateTimeTransformer for Scikit-Learn Datetime Vectorizer

### PR DESCRIPTION
**Pull request by yunma10**

### CODE REVIEW

File: src/dlr_data_transform.cc [hunk 1]
```diff
@@ -126,13 +125,127 @@ void CategoricalStringTransformer::MapToNDArray(const nlohmann::json& input_json
...
 }
 
+tvm::runtime::NDArray DateTimeTransformer::InitNDArray(const nlohmann::json& input_json,
+                                                       DLDataType dtype, DLContext ctx) const {
+  // Create NDArray for transformed input which will be passed to TVM. NUM_COL
+  // fixed to 7
+  std::vector<int64_t> arr_shape = {static_cast<int64_t>(input_json.size()),
+                                    static_cast<int64_t>(kNumDateTimeCols)};
+  CHECK(dtype.code == kDLFloat && dtype.bits == 32 && dtype.lanes == 1)
+      << "DataTransform CategoricalString is only supported for float32 "
```
Comment (trevor-m) [inline src/dlr_data_transform.cc:135]: Change `CategoricalString`

File: src/dlr_data_transform.cc [hunk 2]
```diff
@@ -126,13 +125,127 @@ void CategoricalStringTransformer::MapToNDArray(const nlohmann::json& input_json
...
+  datetime_digits[5] = month;
+  datetime_digits[6] = week_of_year;
+}
+
+void DateTimeTransformer::MapToNDArray(const nlohmann::json& input_json,
+                                       const nlohmann::json& transform,
+                                       tvm::runtime::NDArray& input_array) const {
+  DLTensor* input_tensor = const_cast<DLTensor*>(input_array.operator->());
+  CHECK_EQ(input_tensor->ctx.device_type, DLDeviceType::kDLCPU)
+      << "DataTransform CategoricalString is only supported for CPU.";
```
Comment (trevor-m) [inline src/dlr_data_transform.cc:227]: Change CategoricalString

File: src/dlr_data_transform.cc [hunk 3]
```diff
@@ -126,13 +125,127 @@ void CategoricalStringTransformer::MapToNDArray(const nlohmann::json& input_json
...
+  float* data = static_cast<float*>(input_tensor->data);
+
+  std::vector<int64_t> datetime_digits = std::vector<int64_t>(kNumDateTimeCols, 0);
+  for (size_t r = 0; r < input_json.size(); ++r) {
+    std::string entry = input_json[r][0].get_ref<const std::string&>();
+    DigitizeDateTime(entry, datetime_digits);
+    for (size_t c = 0; c < kNumDateTimeCols; ++c) {
+      const int out_index = r * kNumDateTimeCols + c;
+      data[out_index] = (float)datetime_digits[c];
```
Comment (trevor-m) [inline src/dlr_data_transform.cc:236]: Let's use `static_cast<float>`

File: tests/cpp/dlr_data_transform_test.cc [hunk 4]
```diff
@@ -147,6 +147,42 @@ TEST(DLR, DataTransformMultipleColumn) {
...
+            {
+              "Type": "DateTime"
+            }
+          ]
+        }
+      }
+    })"_json;
+  EXPECT_TRUE(transform.HasInputTransform(metadata));
+
+  const char* data = R"([["Jan 3rd, 2018, 1:34am"], ["Feb 11th, 2012, 11:34:59pm"]])";
```
Comment (trevor-m) [inline tests/cpp/dlr_data_transform_test.cc:166]: Could you add some bad entries here to test that they are handled gracefully? Maybe like some empty input, just a number, a random word.

**Review by trevor-m — CHANGES_REQUESTED** [review]
Thanks @yunma10 just a few more small suggestions

**Review by trevor-m — APPROVED** [review]
Thanks @yunma10 !
